identity: add spire identity client

zaharidichev opened the pull request:
This PR adds SPIRE identity client that allows us to connect to the SPIRE workload API and stream new certificates through the `identity::Credentials` API. Notable changes include: 

- added `spire-proto` crate that includes the workload API protobuf definitions and codegen utilities
- added `spire-client` crate that contains all the facilities to connect to the SPIRE UDS socket and consume updates
- moved the Identity metrics into a seprate crate so it can be reused across both the control plane and spire identity clients

TODO: 
 - add more tests
 - validate and use trust roots received from SPIRE

Signed-off-by: Zahari Dichev <zaharidichev@gmail.com>

mateiidavid commented on linkerd/proxy/spire-client/src/lib.rs:
```diff
@@ -0,0 +1,244 @@
+    {
+        match new_client.call(()).await {
+            Ok(rx) => consume_updates(&self.id, rx, credentials, &self.metrics).await,
+            Err(error) => error!("could not establish SVID stream: {}", error.into()),
```
I wonder if it's better to just have the error message as a field:
```
error!(%error, "could not establish SVID stream");
```

mateiidavid commented on linkerd/proxy/spire-client/src/api.rs:
```diff
@@ -0,0 +1,151 @@
+
+        Ok(Svid {
+            spiffe_id,
+            leaf: leaf.clone(),
```
Do we need the clone here? Can we just move the leaf into the Svid struct? Don't think we're going to be sharing the leaf anywhere else outside of this block.

mateiidavid commented on linkerd/proxy/spire-client/src/client.rs:
```diff
@@ -0,0 +1,63 @@
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    socket: Arc<String>,
```
This is a very nitpicky comment but this feels more like a socket addr rather than a socket (which I'd tend to think more of as a fd). 🤷🏻

mateiidavid commented on linkerd/proxy/spire-client/src/api.rs:
```diff
@@ -0,0 +1,151 @@
+use tower::Service;
+use tracing::error;
+
+#[derive(Clone, Debug)]
```
Should we remove this Debug impl since we're holding the private key bytes?

mateiidavid reviewed (commented):
I know this is still marked as draft but I had a look just to understand the changes. I won't be in for a while after today. Left some drive-by comments.

zaharidichev commented:
@olix0r 

I have: 
- moved the GRPC client into app/src as we discussed
- added more tests
- wired the Spire task into the identity without externalizing these capabilities through the config

olix0r commented on linkerd/app/src/identity.rs:
```diff
@@ -66,33 +83,52 @@ impl Config {
-
-        let addr = self.control.addr.clone();
-
+        let metrics = IdentityMetrics::default();
```
this seems like we're not actually wiring up metrics here.

olix0r commented on linkerd/app/src/identity.rs:
```diff
@@ -30,8 +41,14 @@ pub struct TlsParams {
     pub trust_anchors_pem: String,
 }
 
+#[derive(Clone)]
+pub enum Addr {
+    Spire(Arc<String>),
+    Linkerd(control::ControlAddr),
+}
```
Is it possible to move this into the Provider type to avoid having redundant enums?

zaharidichev replied:
This is used in the `App`, [here](https://github.com/linkerd/linkerd2-proxy/pull/2580/files#diff-4689f302ab1ea17c575cc6e99ad1906a82f3fa3d387b945fe3c25567a6f5635aR351). It was returning a  `control::ControlAddr` before.

olix0r commented on linkerd/proxy/identity-client-metrics/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "linkerd-proxy-identity-client-metrics"
```
Here's an alternative approach to this (against main) https://github.com/linkerd/linkerd2-proxy/pull/2617 -- updating these metrics to reflect best practices (e.g. _total vs _count); it also means neither clients need to be explicitly aware of these metrics.

olix0r commented on linkerd/proxy/spire-client/src/lib.rs:
```diff
@@ -0,0 +1,284 @@
+        S: Service<(), Response = tonic::Response<watch::Receiver<SvidUpdate>>>,
+        S::Error: Into<Error> + Display,
+    {
+        match client.call(()).await {
```
I'd like to talk a little bit more about the error handling in this task, but to satisfy the Service contract, we need to await readiness with something like `client.ready().await?.call(().await` (using the ServiceExt trait to get the ready() call).

olix0r commented on linkerd/proxy/spire-client/src/lib.rs:
```diff
@@ -0,0 +1,284 @@
+        return credentials.set_certificate(svid.leaf, svid.intermediates, svid.private_key, exp);
+    }
+
+    Err("could not find an SVID".into())
```
We tend not to use string errors, instead using typed errors via the thiserror crate.
```rust

#[derive(Debug, thiserror::Error)]
#[error("no matching SVID found")]
pub struct NoMatchingSVIDFound(());
```

Note that the unit field ensures the error can't be built outside of the module.

```suggestion
    Err(NoMatchingSVIDFound(()).into())
```

olix0r commented on linkerd/proxy/spire-client/src/lib.rs:
```diff
@@ -0,0 +1,284 @@
+{
+    loop {
+        let svid_update = updates.borrow_and_update().clone();
+        if let Err(error) = process_svid(&mut credentials, svid_update, id) {
+            tracing::error!("Error processing SVID update: {}", error);
+        }
+
+        if let Err(error) = updates.changed().await {
+            tracing::debug!("SVID watch closed; terminating {}", error);
+            return;
+        }
```
nit:

```suggestion
        if let Err(error) = process_svid(&mut credentials, svid_update, id) {
            tracing::error!(%error, "Error processing SVID update");
        }

        if updates.changed().await.is_err() {
            tracing::debug!("SVID watch closed; terminating");
            return;
        }
```

We generally use field interpolation for errors (rather than direct string formatting).

Also, the error type from changed() is uninteresting -- it can only indicate that all receivers were dropped.

olix0r commented on linkerd/proxy/spire-client/src/lib.rs:
```diff
@@ -0,0 +1,284 @@
+        Self { id }
+    }
+
+    pub async fn run<C, S>(self, credentials: C, mut client: S)
```
How should run handle errors?  Our options are basically (1) panic or (2) bubble them up. Given the choice, it's _gnerally_ preferable to bubble them up and, if necessary, panic closer to the root of the callstack.

olix0r commented on linkerd/proxy/spire-client/src/api.rs:
```diff
@@ -0,0 +1,208 @@
+pub struct Svid {
+    pub(super) spiffe_id: Id,
+    pub(super) leaf: DerX509,
+    pub(super) private_key: Vec<u8>,
```
This is a bit of yakshave that we can do as a followup, but I think we really ought to have a linkerd_identity::PrivateKey type so we're not passing around sensitive byte arrays everywhere.

Why does the private key need to be crate-visible?

olix0r replied:
Ah, I see this is for process_svid -- what if we moved process_svid into this module and made _that_ `pub(super)` -- this gives us a narrower module boundary where we can make all of the Svid stuff private

olix0r commented on linkerd/app/src/identity.rs:
```diff
@@ -55,47 +62,85 @@ impl Config {
+                    (Id::Dns(id), sni) if id == sni => id.clone(),
+                    (_id, _sni) => {
+                        return Err(
+                            "Linkerd identity requires a TLS Id and server name to be the same"
```
Ideally we'd use thiserror again here. Since is up at the app layer I don't care as much as I would in a lib.

olix0r commented on linkerd/proxy/spire-client/src/lib.rs:
```diff
@@ -0,0 +1,257 @@
+        S: Service<(), Response = tonic::Response<watch::Receiver<SvidUpdate>>>,
+        S::Error: Into<Error> + Display + Debug,
+    {
+        let client = client.ready().await.expect("should be ready");
+        match client.call(()).await {
+            Ok(rsp) => consume_updates(&self.id, rsp.into_inner(), credentials).await,
+            Err(error) => error!(%error, "could not establish SVID stream"),
+        }
```
```suggestion
        let client = client.ready().await.expect("should be ready");
        let rsp = client.call(()).await.expect("spire client must gracefully handle errors");
        consume_updates(&self.id, rsp.into_inner(), credentials).await
```

Based on our earlier convo, I'm pretty sure that a failure here indicates something pathological (i.e., a bad request from the client).  If so, it seems better to panic than to log and never satisfy an identity.

olix0r reviewed (commented):
LGTM. One last (I think ;) nit

olix0r reviewed (approved):
:taco:

olix0r commented on spiffe-proto/src/gen/spiffe.workloadapi.rs:
```diff
@@ -0,0 +1,151 @@
+#[allow(clippy::derive_partial_eq_without_eq)]
+#[derive(Clone, PartialEq, ::prost::Message)]
+pub struct X509svidRequest {}
+/// The X509SVIDResponse message carries X.509-SVIDs and related information,
```
```suggestion

/// The X509SVIDResponse message carries X.509-SVIDs and related information,
```
